BUGFIX: Support PHP8 union and intersection types as method return value in reflection service

mficzel opened the pull request:
The change allows to use union types and intersection types in classes which are proxied by the flow framework. 
Offcourse the php version Flow is running on has to support the used features aswell.

In addition: 
- `true`, `false` and `null` are added to the list of simple types which are not prefixed by the reflection service
- `parent`, `static`, `never` and `resource` are added to the list of additional types that need no prefix during reflection aswell

Resolves: #2941 

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked wit `!!!` and have upgrade-instructions

mficzel committed:
TASK: Add functional test for failing handling of union types in method returns

mficzel committed:
TASK: Add "isLiteralType" "isUnionType" and "isIntersectionType" to the TypeHandling class

mficzel committed:
TASK: Adjust reflection service to avoid prefixing literal types with a / and respect union types in return types

mficzel commented:
@albe do you agree with this solution in principle? What should we tests and what branch to target? Is this a bugfix? I personally think this should be a bugfix for 8.0.

Also do we need to handle literal, union and intersection types in method arguments and property types aswell?

albe commented on Neos.Utility.ObjectHandling/Classes/TypeHandling.php:
```diff
@@ -114,6 +114,17 @@ public static function isSimpleType(string $type): bool
+     */
+    public static function isLiteralType(string $type): bool
+    {
+        return in_array(self::normalizeType($type), ['false', 'true'], true);
```
```suggestion
        return in_array(self::normalizeType($type), ['false'], true);
```
I think `true` is not a valid literal type, only `false` is allowed. `null` would be another value in the same category (but we could discuss that in another PR). https://php.watch/versions/8.2/null-false-types

mficzel replied:
true is valid since php 8.2 that is why i included it. See https://www.php.net/releases/8.2/en.php#null_false_true_types

mficzel replied:
Added `null` to the simpleTypes as it is documented here https://www.php.net/manual/de/language.types.type-system.php

mficzel replied:
Also: I wonder wether we should simply add `true` and `false` to the list of simple types and avoid the seperate "isLiteralType" method. Not sure.

albe commented on Neos.Flow/Classes/Reflection/ReflectionService.php:
```diff
@@ -1447,12 +1447,27 @@ protected function reflectClassMethod($className, MethodReflection $method)
+            return $type;
+        };
+        if ($returnType !== null) {
+            if (TypeHandling::isUnionType($returnType)) {
```
Basically this is okay as is for now.
I do think we should move all this type handling stuff back into PHP standard reflection API land though. So instead of our custom `getDeclaredReturnType()` that returns the stringified type, then parse it back if it was a union/intersection type, we should use the `getReturnType()` method as primary source of information directly
https://www.php.net/manual/en/class.reflectiontype.php
That can quickly become complex, because we might want to wrap that into a new method that falls back to docblock parsing info, which would mean we'd need to use own inherited `ReturnType` classes that we can manually instanciate (I think we don't do this yet, because we do the fallback way deeper in the code, but actually the ReflectionService should be the place to "normalize" PHP type declarations and docblock types into one souce of information). Or maybe there is a simpler intermediate step.

mficzel replied:
Maybe we can parse the docblocks to ReflectionTypes ...

mficzel replied:
Reading again parsing docblocks to reflections is exactly the core of your suggestion. So i agree to that.

albe reviewed (approved):
I think this is okay for a bugfix solution to get union+intersection types to work. I outlined my prefered (midterm) solution in the comment.

> Also do we need to handle literal, union and intersection types in method arguments and property types aswell?

For completeness yes, but I'm not even sure where/how we do reflect on those in a specific way (we have https://github.com/mficzel/flow-development-collection/blob/bugfix/handleUnionTypes/Neos.Flow/Classes/Reflection/ParameterReflection.php but that does not do anything fancy besides the `isBuiltin`), that requires adjustment. A lot of adjustments will also be necessary in higher levels, where we do use the Reflection info, and that's a way bigger topic to tackle.

mficzel committed:
TASK: Add `null` type to simple types.

mficzel commented:
Bugfix for 8.0 or 7.3? Older versions without php 8 support are not affected imho.

fcool commented:
I would definitely vote for Bugfix in 7.3

fcool commented on Neos.Flow/Tests/Functional/Reflection/Fixtures/DummyClassWithTypeHints.php:
```diff
@@ -31,4 +31,8 @@ public function methodWithArrayTypeHint(array $array)
     {
     }
+
+    public function methodWithUnionReturnType(): string|false
```
This in a test like that will probably break Testing for older PHP versions, wouldn't it? I guess we need some kind of language level detection to skip tests, which would break older PHP parsers - but I am not 100% sure.

The current Tasks only checked PHP >=8.0, if we were targetting 7.3 it would need to degrade gracefully in PHP 7.4

fcool replied:
(If we stay with 8.x as target, that probably is not neccessary - and I'm fine with all other aspects.)

mficzel replied:
That is why i would prefer to target 8.0. Support 7.3 would require some additional adjustments.

mficzel committed:
TASK: Extract php 8 specific test classes into separate folder for Neos 7.3 compatibility

mficzel committed:
TASK: Remove unneeded type normalization in isLiteralType

mficzel commented:
@fcool @albe adjusted and rebased to 7.3 .... could actually be ready. 

The psalm errors are caused by changes in doctrine ... again.

mficzel commented:
I wonder wether the use of "str_contains" (php8) is ok. There is a polyfill from symfony that is loaded as dependency of the symfony components anyways. Also i tested on Neos 7.3 &  php 7.4 in addition to the test matrix here.

grebaldi commented on Neos.Utility.ObjectHandling/Classes/TypeHandling.php:
```diff
@@ -111,7 +111,18 @@ public static function isLiteral(string $type): bool
     public static function isSimpleType(string $type): bool
     {
-        return in_array(self::normalizeType($type), ['array', 'string', 'float', 'integer', 'boolean'], true);
+        return in_array(self::normalizeType($type), ['array', 'string', 'float', 'integer', 'boolean', 'null'], true);
+    }
+
+    /**
+     * Returns true if the $type is a literal type.
+     *
+     * @param string $type
+     * @return boolean
+     */
+    public static function isLiteralType(string $type): bool
+    {
+        return in_array($type, ['false', 'true'], true);
```
I'm confused by the semantics here. The reason is basically that I don't know what the idea of "simple type" (which has not been introduced in this PR) refers to.

There are three methods `isLiteral`, `isSimpleType` and `isLiteralType`. Looking at:
https://github.com/neos/flow-development-collection/blob/7d02951763753b9ff80bb6b4a478b20631e76059/Neos.Utility.ObjectHandling/Classes/TypeHandling.php#L29-L32

I assume that `isLiteral` and `isSimpleType` address the same set of types (in that case, the `LITERAL_TYPE_PATTERN` would need to be adjusted to cover `null` as well).

But I'd like to suggest another way:

Maybe this would be a good opportunity to adjust the `TypeHandling` class to operate closer to PHP's type system definitions (as per https://www.php.net/manual/en/language.types.type-system.php). I'd suggest the following methods:

* `isScalarType`
  * covers `bool`, `int`, `float` and `string`
* `isBuiltInType`
  * covers `null`, `array`, `object`, `resource`, `never`, `void`, `self`, `parent` and `static`
  * falls back to `isScalarType`
* `isLiteralType`
  * Same as it is right now in this PR
* `isCallableType`
  * covers `callable`
* `isBuiltInTypeAlias`
  * covers `mixed` and `iterable`
* `isUserDefinedType`
  * calculated from: `!isBuiltInType && !isLiteralType && !isCallableType && !isBuiltInTypeAlias`
  * can be used in `ReflectionService` to determine whether the type name needs to be prefixed with `\` or not

For backwards compatibility `isLiteral` and `isSimpleType` need to continue to work, but can use the new methods. Given my suggestion, `isLiteral` and `isSimpleType` should (probably) also be deprecated.

mficzel replied:
@grebaldi the idea of literal type comes from php itself https://www.php.net/manual/de/language.types.type-system.php but I find it weird myself and would be fine with just adding true false null to the simple types.

grebaldi commented on Neos.Flow/Classes/Reflection/ReflectionService.php:
```diff
@@ -1410,12 +1410,27 @@ protected function reflectClassMethod($className, MethodReflection $method)
-            $returnType = '?' . $returnType;
+        $returnType= $method->getDeclaredReturnType();
+        $applyLeadingSlashIfNeeded = function (string $type): string {
+            if (!in_array($type, ['self', 'null', 'callable', 'void', 'iterable', 'object', 'mixed'])
+                && !TypeHandling::isSimpleType($type)
+                && !TypeHandling::isLiteralType($type)
```
The type range covered here is not enough. As per https://www.php.net/manual/en/language.types.type-system.php, `resource`, `never`, `parent` and `static` must also be covered.

Also: I think that L1415 should be encapsuled in `TypeHandling` as well (namely through the newly suggested `isUserDefinedType` method).

mficzel replied:
Makes sense. Never and void still need to be handled separately as they cannot be used in union types. 

Also I just noticed that union types can contain intersection types in brackets … but it would be fair to ignore this in a bugfix for Neos 7.3

grebaldi reviewed (changes requested):
Hi @mficzel,

I was able to reproduce the issue and to verify that your change works as expected. While browsing the Code however, I saw some odd semantics in the `TypeHandling` class and thus discovered that the original bug has a whole other dimension :scream: 

Long story short: There are still PHP return type names that would be destroyed during Proxy Class building (I was able to create such a class). I added some code comments to describe the issue more thoroughly.

mficzel committed:
TASK: Move true and false to the simple types and remove the method isLiteralType as it was confusing

mficzel committed:
TASK: add missing builtin types `resource`, `never`, `parent`, `static` and `resource` to the list of types that do not need prefixing

mficzel commented:
@grebaldi, thank you for your detailed feedback. I think i covered the cases you found: 

- I moved `true` and `false` to the isSimpleType method as i agree that the semantic of isLiteral vs isLiteralType is bad. As SimpleTypes are no concept of PHP we can aswell just add the values there.
- I added the missing exceptions for `static`, `parent`, `never` and `resource` to the exception list in the ReflectionService

I 100% agree to your suggestions for the additional methods in TypeHandling but opted to not introduce those in a bugfix but would add those rather as a feature to 8,3 after this fix was upmerged.

Also i did not handle the cases when never or void are wrongly used in a union type specially as it is correct that the reflection does not modify those types and php will throw the correct errors anyways. So i consider the current behavior in this regard already ok.

What also is not handled yet is nesting of intersection types inside union types. I consider this too much for this bugfix. This should be solved with a feature for a Neos release.

grebaldi reviewed (approved):
Hi @mficzel,

thanks for your adjustments! LGTM :)

> I 100% agree to your suggestions for the additional methods in TypeHandling but opted to not introduce those in a bugfix but would add those rather as a feature to 8,3 after this fix was upmerged.

I opened an issue to cover this: https://github.com/neos/flow-development-collection/issues/2945

mficzel commented:
@robertlemke @albe @fcool a second approval would be great

fcool commented:
Will definitely test and check it at latest tomorrow morning!

fcool reviewed (approved):
Just a comment: Flow is still broken, if you use a type hint of `object` in the signature of a proxied method. :(

Error is: `PHP Fatal error:  Type declaration 'object' must be unqualified in ...`

But this is probably out of scope of this Bugfix, so I'm fine with creating a new issue for this behavior.

mficzel commented:
Just noticed that almost the same problems occur when union types or other types flow does not yet handle correctly are used in method parameters ... see https://github.com/neos/flow-development-collection/blob/7.3/Neos.Flow/Classes/Reflection/ReflectionService.php#L1834 

I wonder how to approach this, probably #2945 is needed for that.

albe commented:
> I wonder how to approach this, probably #2945 is needed for that.

Yeah, anything beyond the (incomplete) scope of this bugfix will need much deeper change. So as long as this change doesn't build up more stumbling blocks for that bigger change (which I don't see), we should get it in IMO.

mficzel committed:
TASK: Update psalm baseline
